src/Land/Model/source.jl

```diff
@@ -82,35 +69,21 @@ function land_source!(
             heaviside(_Tfreeze - T) *
             heaviside(θ_l - θstar) - _ρice * θ_i * heaviside(T - _Tfreeze)
         )
-    source.soil.water.ϑ_l -= freeze_thaw / _ρliq
-    source.soil.water.θ_i += freeze_thaw / _ρice
+    return (; freeze_thaw)
 end
 
+function source(::ϑLiquid, s::PhaseChange, land::LandModel, args)
+    @unpack state = args
+    @unpack freeze_thaw = args.precomputed.dtup[s]
+    FT = eltype(state)
+    _ρliq = FT(ρ_cloud_liq(land.param_set))
+    return -freeze_thaw / _ρliq
+end
 
-# sources are applied additively
-
-@generated function land_source!(
-    stuple::Tuple,
-    land::LandModel,
-    source::Vars,
-    state::Vars,
-    diffusive::Vars,
-    aux::Vars,
-    t::Real,
-    direction,
-)
-    N = fieldcount(stuple)
-    return quote
-        Base.Cartesian.@nexprs $N i -> land_source!(
-            stuple[i],
-            land,
-            source,
-            state,
-            diffusive,
-            aux,
-            t,
-            direction,
-        )
-        return nothing
-    end
+function source(::θIce, s::PhaseChange, land::LandModel, args)
+    @unpack state = args
+    @unpack freeze_thaw = args.precomputed.dtup[s]
+    FT = eltype(state)
+    _ρice = FT(ρ_cloud_ice(land.param_set))
```

Comment:
`p_cloud_liq`

Reply:
this one is OK - we want rho_i d(\theta_i)/dt = - rho_l d(\theta_l)/dt . theta is proportional to the volume in the phase, so this makes it so freezing/thawing conserves the mass of water, while the volume changes by the right amount.